src/core/src/op/loop.cpp

```diff
@@ -191,6 +191,66 @@ void op::v5::Loop::validate_and_infer_types() {
+            back_edges[merged_desc->m_body_value_index] = merged_desc->m_body_parameter_index;
+        }
+    }
+    if (!back_edges.empty()) {
+        bool need_reinvalidate = false;
+        for (const auto& output_description : m_output_descriptions[0]) {
+            auto body_value = m_bodies[0]->get_results().at(output_description->m_body_value_index)->input_value(0);
+
+            if (auto body_output_description =
+                    ov::as_type_ptr<v0::TensorIterator::BodyOutputDescription>(output_description)) {
+                const ov::PartialShape& body_value_shape = body_value.get_partial_shape();
+                if (body_value_shape.rank().is_static()) {
+                    // handle the case: when sub-model's output shape does not compatible input shape in a back-edge,
+                    // such as
+                    //          Parameter(out:-1, 1)->|
+                    //                                |->Concat(out:-1, 2)->Result(out:-1, 2)
+                    //          Parameter(out:-1, 1)->|   (axis==1)
+                    // when iteration number is unknown or sub-model output shape may be vary, the Result shape should
+                    // infer as (-1, -1), then set changed one to input and propagate to others.
+                    if (back_edges.count(output_description->m_body_value_index)) {
+                        auto input_param =
+                            m_bodies[0]->get_parameters().at(back_edges[output_description->m_body_value_index]);
+                        const auto& input_param_ps = input_param->get_partial_shape();
+                        if (input_param_ps.rank().is_static()) {
+                            const auto body_rank_len = body_value_shape.rank().get_length();
+                            const auto input_rank_len = input_param_ps.rank().get_length();
+                            // if input and output shape are {n} or {} and not compatible, make output shape dynamic
+                            if (body_rank_len <= 1 && input_rank_len <= 1 &&
+                                !input_param_ps.compatible(body_value_shape)) {
+                                // reset sub model input shape
+                                input_param->set_partial_shape({-1});
+                                need_reinvalidate = true;
+                            } else if (body_rank_len == input_rank_len &&
+                                       !input_param_ps.compatible(body_value_shape)) {
+                                ov::PartialShape new_ps(body_value_shape);
+                                for (auto i = 0; i < body_value_shape.size(); i++) {
+                                    if (!body_value_shape[i].compatible(input_param_ps[i])) {
+                                        new_ps[i] = Dimension::dynamic();
+                                    }
+                                }
```

Comment:
Unfortunately, it breaks combination transformations: RemoveConcatZeroDimInput and RemoveMultiSubGraphOpDanglingParams which enable important model from 57997. As a result danling Loop's body subgraph that contains a tensor with zero dimension (which means also zero allocated memory) is not removed and invalid pointer exception is thrown.
I am aware that it is the most elegant solution, but we probably should add additonal if for such case. It can be sth like:
```
for (auto i = 0; i < body_value_shape.size(); i++) {
	if (!body_value_shape[i].compatible(input_param_ps[i])) {
		if (input_param_ps[i].is_static() && input_param_ps[i].get_length() == 0) {
			new_ps[i] = input_param_ps[i];
		} else {
			new_ps[i] = Dimension::dynamic();
		}
	}
}
```

Reply:
Unfortunately the zero length input does exist in paddle faster-rcnn model just like the following:
```
main model: Const(1, 0, 4) -> Loop -> Result(1, -1, 4)
sub model: Parameter(1, -1, 4) -> Concat(axis=1) -> Result(1, -1, 4)
         Const(1,3,4)          -> |
```
I did a test based on the PR and showed that concat zero length shape node could work. The test script:
```python
from openvino.runtime import Core, Model, Tensor, opset8 as opset
from openvino.runtime.passes import Manager
import numpy as np

# simulate the following code(tensor array initial shape in paddle model 
#   always has a zero dim)
# result = []
# c = [0] * 10
# for _ in range(4)
#   result.append(c)           # ir will use concat instead
# assert(len(result) == 4)

core = Core()
# zero dim constant
init_0_dim = opset.constant(np.empty(shape=(0, 10)), np.float32, name='init 0 dim array')
trip = opset.parameter(dtype=np.int64, shape=[1, ], name='trip')

# make a sub model
cond = opset.constant(True, dtype=np.bool, name='cond')
# zero dim parameter, the first concat input shapes should be [0,10] and [1,10]
# it works
sub_x = opset.parameter(dtype=np.float32, shape=[0, 10], name='x')
sub_cond = opset.parameter(dtype=np.bool, shape=[1], name='cond')
sub_c = opset.constant(np.zeros(shape=(1, 10)), np.float32)
sub_concat = opset.concat([sub_x, sub_c], 0)
sub_result_x = opset.result(sub_concat)
sub_result_cond = opset.result(sub_cond)
sub_model = Model([sub_result_x, sub_result_cond], [sub_x, sub_cond], name='submodel')

loop = opset.loop(trip, cond)
loop.set_function(sub_model)
# loop back: concat result will be the input of the 'x'
loop.set_merged_input(sub_x, init_0_dim.output(0), sub_result_x.output(0))
loop.set_invariant_input(sub_cond, cond.output(0))
loop.set_special_body_ports([-1, 1])
out0 = loop.get_iter_value(sub_result_x.output(0), -1)
out2 = loop.get_iter_value(sub_result_cond.output(0), -1)

# main model
m = Model([out0, ], [trip, ], "test")
net = core.compile_model(m, 'CPU')
req = net.create_infer_request()
number = 4
req.set_tensors({'trip': Tensor(np.array([number], dtype=np.int64)),})
result = req.infer()
assert(req.outputs[0].data.shape[0] == number)

model = net.get_runtime_model()
man = Manager()
man.register_pass("Serialize", output_files=(f"pp.xml", f"pp.bin"))
man.run_passes(m)
```

Reply:
@luo-cheng2021 I see your point, but anyway empty tensor will ignored during execution by a plugin. What's more the equivalent transformation is also implemented in MO https://github.com/openvinotoolkit/openvino/blob/fd97a622635f4cc85d4efce1a5fed4d9f2fa2246/model-optimizer/extensions/middle/CutInputHavingZeroDimFromConcat.py#L9.
In addition, calculation subgraph which doesn't provide any information seems to be unuseful.

Does the proposed change break paddle faster-rcnn model even when `RemoveConcatZeroDimInput` is turned off via `disable_remove_concat_zerodim_input` ?

Also, I've checked that if https://github.com/openvinotoolkit/openvino/pull/11547/files will be introduced and 0 dim will be passed to body params as suggested, transformations `RemoveConcatZeroDimInput` and `RemoveMultiSubGraphOpDanglingParams` can be removed from `openvino/src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp` and be kept only in `CNNNetworkNGraphImpl::CNNNetworkNGraphImpl`. Does it change anything for you?

Reply:
> In addition, calculation subgraph which doesn't provide any information seems to be unuseful.

Although the subgraph seams useless itself, it will produce a useful result after 1 iteration.

> Does the proposed change break paddle faster-rcnn model even when `RemoveConcatZeroDimInput` is turned off via `disable_remove_concat_zerodim_input` ?

Yes, it does break. We need the zero dimension constant in paddle faster-rcnn model.
 
> 
> Also, I've checked that if https://github.com/openvinotoolkit/openvino/pull/11547/files will be introduced and 0 dim will be passed to body params as suggested, transformations `RemoveConcatZeroDimInput` and `RemoveMultiSubGraphOpDanglingParams` can be removed from `openvino/src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp` and be kept only in `CNNNetworkNGraphImpl::CNNNetworkNGraphImpl`. Does it change anything for you?

I tested the changes in my PR and it's ok to me.

Reply:
Unfortunetelly, my idea to remove `RemoveConcatZeroDimInput ` from `moc_transformations.cpp` doesn't work on all cases (for example via ModelOptimizer).

Just to show you my case:
![image](https://user-images.githubusercontent.com/11331953/169974778-b61b2bfd-9e34-4075-8910-0deba0b9101c.png)

In your proposition we are losing information about body value shape (by setting `Dimension::dynamic`) so `RemoveConcatZeroDimInput` doesn't work.

I've chcecked paddle faster-rcnn model and it works **without** such change.

> Although the subgraph seams useless itself, it will produce a useful result after 1 iteration.
Could you clarify this point? As I checked, for example CPU plugin ignore such tensor in current state: https://github.com/openvinotoolkit/openvino/blob/22ee17fda6ba44c05c537d309923557eb1febd1c/src/plugins/intel_cpu/src/nodes/concat.cpp#L507

@luo-cheng2021  Do you maybe have some idea to compromise here?

Reply:
> Unfortunetelly, my idea to remove `RemoveConcatZeroDimInput ` from `moc_transformations.cpp` doesn't work on all cases (for example via ModelOptimizer).

I did not test MO yet, if needed I think I need another PR to fix MO part. Thanks for reminding.

> I've chcecked paddle faster-rcnn model and it works **without** such change.

Without such change faster-rcnn will produce fix length(zero or one) nms result, it could run but has wrong results.

> > Although the subgraph seams useless itself, it will produce a useful result after 1 iteration.
> > Could you clarify this point? As I checked, for example CPU plugin ignore such tensor in current state: https://github.com/openvinotoolkit/openvino/blob/22ee17fda6ba44c05c537d309923557eb1febd1c/src/plugins/intel_cpu/src/nodes/concat.cpp#L507
> 
> @luo-cheng2021 Do you maybe have some idea to compromise here?

The `srcMem.GetShape()` will be different value with the previous in the next iteration, the input shape will not be zero after the first iteration. In cpu plugin [TensorIterator](https://github.com/openvinotoolkit/openvino/blob/5b75d69712042288315d3c24ffb8e34195e1b773/src/plugins/intel_cpu/src/nodes/tensoriterator.cpp#L566) will copy the output to the input using [prepareDynamicBackEdges](https://github.com/openvinotoolkit/openvino/blob/5b75d69712042288315d3c24ffb8e34195e1b773/src/plugins/intel_cpu/src/nodes/tensoriterator.cpp#L611) if dynamic shape, prepareDynamicBackEdges will call [redefineToMemories](https://github.com/openvinotoolkit/openvino/blob/5b75d69712042288315d3c24ffb8e34195e1b773/src/plugins/intel_cpu/src/nodes/tensoriterator.cpp#L60) which may changes the shapes that the inputs will use in the next iteration.

Reply:
@luo-cheng2021 I've prepared addional transformation (https://github.com/openvinotoolkit/openvino/pull/11957) which works for onnx models and also doesn't break case with empty tensor/Concat in Loop (in combined your PR + my change). Please for review